### PR TITLE
Support new PAL metadata layout on graphics pipeline in middle-end

### PR DESCRIPTION
**Pull request by xuechen417**
The old PAL metadata for GFX9-plus is a 32-bit interger value for each register
```
.registers {
   PA_SU_VTX_CNTL        0x000000000000002D
      ...}
 ```
 The new PAL metadata layout for GFX9-plus is field-value pairs for each
 register
 ```
 .registers {
     .graphics_registers {
          .pa_su_vtx_cntl: {
             .pix_center: 1
	     .round_mode: 0x0000000000000002
	     .quant_mode: 0x0000000000000}
...}}}
```
This change will build the framework to support the register field format for GFX9-plus in middle-end.
1. The new derived class `RegisterMetadaBuilder` is to generate the new metadata layout. `buildPalMetatdata` calls `build{LsHs,EsGS,Prim,HwVs,Ps}Registers` methods to set the registers.
2. The new cl::option `UseRegisterFieldFormat` is to control whether employing the new PAL metadata layout and it is turned off by default.

### CODE REVIEW

**Comment by xuechen417** [comment]
V1: Fix typos

**Comment by amdvlk-admin** [comment]
### Test summary for commit af4ab07fd29b5c42ff62a95cbb7c45bcb3d4e2db
<details><summary>CTS tests (Failed: 0/283117)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38864/70780 (54.9%)</li>
<li>Failed: 0/70780 (0.0%)</li>
<li>Not Supported: 31916/70780 (45.1%)</li><li>Warnings: 0/70780 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 39114/70780 (55.3%)</li>
<li>Failed: 0/70780 (0.0%)</li>
<li>Not Supported: 31666/70780 (44.7%)</li><li>Warnings: 0/70780 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35959/70782 (50.8%)</li>
<li>Failed: 0/70782 (0.0%)</li>
<li>Not Supported: 34823/70782 (49.2%)</li><li>Warnings: 0/70782 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 35861/70775 (50.7%)</li>
<li>Failed: 0/70775 (0.0%)</li>
<li>Not Supported: 34914/70775 (49.3%)</li><li>Warnings: 0/70775 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/RegisterMetadataBuilder.cpp [hunk 1]
```diff
@@ -0,0 +1,1362 @@
...
+
+#define DEBUG_TYPE "lgc-register-metadata-builder"
+
+using namespace llvm;
+
+namespace lgc {
+
+namespace Gfx9 {
+
+#include "chip/gfx9/gfx9_plus_merged_enum.h"
```
[the remark this answers is not in the document]
Comment (xuechen417) [reply lgc/patch/RegisterMetadataBuilder.cpp:27]: Okay. How about making your suggested changes in another PR?

File: lgc/patch/RegisterMetadataBuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,1362 @@
...
+  ldsSizeInDwords = alignTo(ldsSizeInDwords, ldsSizeDwordGranularity);
+
+  auto hwShaderNode = getHwShaderNode(Util::Abi::HardwareStage::Gs);
+  hwShaderNode[Util::Abi::HardwareStageMetadataKey::LdsSize] = ldsSizeInDwords * 4;
+  setEsGsLdsSize(calcFactor.esGsLdsSize * 4);
+}
+
+// =====================================================================================================================
+// Builds register configuration for hardware primitive shader.
+void RegisterMetadataBuilder::buildPrimRegisters() {
```
[the remark this answers is not in the document]
Comment (xuechen417) [reply lgc/patch/RegisterMetadataBuilder.cpp:309]: Fixed.

File: lgc/patch/RegisterMetadataBuilder.cpp [hunk 3]
```diff
@@ -0,0 +1,1362 @@
...
+
+  auto hwShaderNode = getHwShaderNode(Util::Abi::HardwareStage::Gs);
+  hwShaderNode[Util::Abi::HardwareStageMetadataKey::LdsSize] = ldsSizeInDwords * 4;
+  if (!m_hasMesh)
+    setEsGsLdsSize(calcFactor.esGsLdsSize * 4);
+}
+
+// =====================================================================================================================
+// Builds register configuration for hardware vertex shader.
+void RegisterMetadataBuilder::buildHwVsRegisters() {
```
[the remark this answers is not in the document]
Comment (xuechen417) [reply lgc/patch/RegisterMetadataBuilder.cpp:567]: @nhaehnle  Is that okay with you?

**Review by xuechen417 — COMMENTED** [review]
V1: Address some format issues.

**Comment by amdvlk-admin** [comment]
### Test summary for commit c42be849c9d1603eb8d93b821658d032c98e5aca
<details><summary>CTS tests (Failed: 0/283117)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38864/70780 (54.9%)</li>
<li>Failed: 0/70780 (0.0%)</li>
<li>Not Supported: 31916/70780 (45.1%)</li><li>Warnings: 0/70780 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 39114/70780 (55.3%)</li>
<li>Failed: 0/70780 (0.0%)</li>
<li>Not Supported: 31666/70780 (44.7%)</li><li>Warnings: 0/70780 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35959/70782 (50.8%)</li>
<li>Failed: 0/70782 (0.0%)</li>
<li>Not Supported: 34823/70782 (49.2%)</li><li>Warnings: 0/70782 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 35861/70775 (50.7%)</li>
<li>Failed: 0/70775 (0.0%)</li>
<li>Not Supported: 34914/70775 (49.3%)</li><li>Warnings: 0/70775 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/RegisterMetadataBuilder.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (nhaehnle) [reply lgc/patch/RegisterMetadataBuilder.cpp:585]: I don't feel very strongly about it, but I slightly prefer with the "Hw" prefix since I think it might help avoid confusion.

File: lgc/include/lgc/state/AbiMetadata.h [hunk 4]
```diff
@@ -66,13 +66,13 @@ enum PipelineType : unsigned {
...
-  Vs,     ///< Hardware VS stage
-  Ps,     ///< Hardware PS stage
-  Cs,     ///< Hardware CS stage
+  Ls = 0, // Hardware LS stage
+  Hs,     // Hardware hS stage
+  Es,     // Hardware ES stage
+  Gs,     // Hardware GS stage
+  Vs,     // Hardware VS stage
+  Ps,     // Hardware PS stage
+  Cs,     // Hardware CS stage
```
Comment (nhaehnle) [inline lgc/include/lgc/state/AbiMetadata.h:75]: Is there a reason for removing the "///<"?

While we don't actually use it particularly well, the idea here is that "///<" initiate a "document comment" that refers to the left-hand-side on the same line. It feels slightly better to keep it, and in any case this is just an unnecessary change.

I think the same applies to some other enums.

File: lgc/patch/RegisterMetadataBuilder.cpp [hunk 5]
```diff
@@ -0,0 +1,1380 @@
...
+  ShaderStage apiStage = apiStage2 != ShaderStageInvalid ? apiStage2 : apiStage1;
+
+  if (m_isNggMode || m_gfxIp.major == 10) {
+    unsigned waveFrontSize = m_pipelineState->getShaderWaveSize(apiStage);
+    hwShaderNode[Util::Abi::HardwareStageMetadataKey::WavefrontSize] = waveFrontSize;
+  }
+
+  if (m_pipelineState->getTargetInfo().getGpuProperty().supportShaderPowerProfiling) {
+    unsigned checksum = setShaderHash(apiStage1);
+    checksum ^= setShaderHash(apiStage2);
```
Comment (nhaehnle) [inline lgc/patch/RegisterMetadataBuilder.cpp:869]: Shouldn't this be guarded by a check that `apiStage2 != StageInvalid`?

**Review by nhaehnle — COMMENTED** [review]
I still have a few minor inline comments, but apart from that it looks good to go.

File: lgc/patch/RegisterMetadataBuilder.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (amdrexu) [reply lgc/patch/RegisterMetadataBuilder.cpp:585]: OK. Let's keep Hw prefix.

File: lgc/include/lgc/state/AbiMetadata.h (hunk 4 above)
Comment (amdrexu) [reply lgc/include/lgc/state/AbiMetadata.h:75]: > Is there a reason for removing the "///<"?

We use the mix of "///<" and "//" in comments. I think it is better for us to clarify the usage style in the contribution doc and coding style doc. Otherwise, this is very confusing for new comers to follow. In PAL coding style, only interface files that are used by clients adopt the doxygen style ///< and other source files keep normal style //.

**Comment by xuechen417** [comment]
V2: Remove `UserSgprMsb` and add guard for checksum

**Review by nhaehnle — APPROVED** [review]
Thanks, LGTM!

File: lgc/patch/RegisterMetadataBuilder.cpp [hunk 6]
```diff
@@ -0,0 +1,1380 @@
...
+  const unsigned itemCount = 4;
+  unsigned gsVertItemSizes[itemCount] = {};
+  unsigned gsVsRingOffsets[itemCount] = {};
+  unsigned gsVsRingOffset = 0;
+  for (unsigned i = 0; i < itemCount; ++i) {
+    gsVertItemSizes[i] = sizeof(unsigned) * gsInOutUsage.gs.outLocCount[i];
+    gsVsRingOffset += gsVertItemSizes[i] * maxVertOut;
+    gsVsRingOffsets[i] = gsVsRingOffset;
+  }
+  getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VgtGsVertItemsize] = gsVertItemSizes;
```
Comment (trenouf) [inline lgc/patch/RegisterMetadataBuilder.cpp:265]: This and the next line give a warning "address of array will always evaluate to true" for me. Do you want to assign the whole array to a msgpack array node? There is no operator= for that; you'll have to do it manually.

File: lgc/patch/RegisterMetadataBuilder.cpp [hunk 7]
```diff
@@ -0,0 +1,1380 @@
...
+      spiShaderGsMeshletExpAlloc[Util::Abi::SpiShaderGsMeshletExpAllocMetadataKey::MaxExpVerts] =
+          meshMode.outputVertices;
+      spiShaderGsMeshletExpAlloc[Util::Abi::SpiShaderGsMeshletExpAllocMetadataKey::MaxExpPrims] =
+          meshMode.outputPrimitives;
+    }
+  } else {
+    maxVertsPerSubgroup = std::min(gsInstPrimsInSubgrp * maxVertOut, NggMaxThreadsPerSubgroup);
+    // VGT_GS_VERT_ITEMSIZE
+    unsigned gsVertItemSizes[4] = {4 * gsInOutUsage.outputMapLocCount, 0, 0, 0};
+    getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VgtGsVertItemsize] = gsVertItemSizes;
```
Comment (trenouf) [inline lgc/patch/RegisterMetadataBuilder.cpp:505]: Another "address of array will always evaluate to true".

File: lgc/patch/RegisterMetadataBuilder.cpp [hunk 8]
```diff
@@ -0,0 +1,1380 @@
...
+
+  // Set some field of SPI_SHADER_PGM_RSRC2_VS
+  getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VsSoEn] = enableXfb;
+  getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VsSoBase0En] = xfbStrides[0] > 0;
+  getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VsSoBase1En] = xfbStrides[1] > 0;
+  getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VsSoBase2En] = xfbStrides[2] > 0;
+  getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VsSoBase3En] = xfbStrides[3] > 0;
+
+  // VGT_STRMOUT_VTX_STRIDE_*
+  getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VgtStrmoutVtxStride0] = xfbStrides[0] / sizeof(unsigned);
```
Comment (trenouf) [inline lgc/patch/RegisterMetadataBuilder.cpp:618]: These four lines give errors "use of overloaded operator= is ambiguous" for me. The right side is unsigned long. Probably need to manually cast to unsigned.

Comment (xuechen417) [reply lgc/patch/RegisterMetadataBuilder.cpp:618]: Thank you for pointing out. I will fix them.

File: lgc/patch/RegisterMetadataBuilder.cpp (hunk 6 above)
Comment (xuechen417) [reply lgc/patch/RegisterMetadataBuilder.cpp:265]: Yes, I made a wrong use for the array node. They are fixed in [d061491](https://github.com/GPUOpen-Drivers/llpc/commit/d0614913041bd72c1f3999d373fb19323ee72b2d)

File: lgc/patch/RegisterMetadataBuilder.cpp (hunk 7 above)
Comment (xuechen417) [reply lgc/patch/RegisterMetadataBuilder.cpp:505]: Fixed in [d061491](https://github.com/GPUOpen-Drivers/llpc/commit/d0614913041bd72c1f3999d373fb19323ee72b2d).

**Comment by jayfoad** [comment]
Hi, I am seeing a different warning from the one Tim mentioned:
```
lgc/patch/RegisterMetadataBuilder.cpp:265:85: warning: address of array 'gsVertItemSizes' will always evaluate to 'true' [-Wpointer-bool-conversion]
lgc/patch/RegisterMetadataBuilder.cpp:266:85: warning: address of array 'gsVsRingOffsets' will always evaluate to 'true' [-Wpointer-bool-conversion]
lgc/patch/RegisterMetadataBuilder.cpp:505:87: warning: address of array 'gsVertItemSizes' will always evaluate to 'true' [-Wpointer-bool-conversion]
```

File: lgc/patch/RegisterMetadataBuilder.cpp (hunk 8 above)
Comment (xuechen417) [reply lgc/patch/RegisterMetadataBuilder.cpp:618]: Fix in PR #2264

**Comment by nhaehnle** [comment]
>  These four lines give errors "use of overloaded operator= is ambiguous" for me. The right side is unsigned long. Probably need to manually cast to unsigned.

Given a similar internal discussion, this seems like a subtle difference in compiler versions. Maybe it's time to look into the compiler versions covered by CI and make a change?
